src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,41 +1,91 @@
 package com.es.phoneshop.web;
 
+import com.es.phoneshop.exceptions.OutOfStockException;
+import com.es.phoneshop.model.cart.Cart;
 import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.service.CartService;
+import com.es.phoneshop.service.impl.CartServiceImpl;
 import com.es.phoneshop.service.ProductService;
-import com.es.phoneshop.service.ProductServiceImpl;
+import com.es.phoneshop.service.RecentlyViewedProductsService;
+import com.es.phoneshop.service.impl.ProductServiceImpl;
+import com.es.phoneshop.service.impl.RecentlyViewedProductsServiceImpl;
 import jakarta.servlet.ServletConfig;
 import jakarta.servlet.ServletException;
 import jakarta.servlet.http.HttpServlet;
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.servlet.http.HttpServletResponse;
+import jakarta.servlet.http.HttpSession;
 
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
 
 public class ProductDetailsPageServlet extends HttpServlet {
 
+    private RecentlyViewedProductsService recentlyViewedProductsService;
     private ProductService productService;
+    private CartService cartService;
 
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productService = ProductServiceImpl.getInstance();
+        cartService = CartServiceImpl.getInstance();
+        recentlyViewedProductsService = RecentlyViewedProductsServiceImpl.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo();
-        if (productId != null && productId.contains("/")) {
-            productId = productId.substring(productId.indexOf('/') + 1);
-        }
-
-        long id = Long.parseLong(productId);
+        long id = parseProductId(request);
         Product product = productService.getProduct(id);
+        HttpSession session = request.getSession();
+
         if (product != null) {
+            recentlyViewedProductsService.addRecentlyViewedProduct(session, product);
             request.setAttribute("product", product);
+            request.setAttribute("cart", cartService.getCart(request));
```

Comment:
cart (same as recentlyViewedProducts) is stored in session
no need to duplicate it to the request